Reduce need to specify input pipeline.

jpienaar opened the pull request:
Utilize dynamic pipeline to construct appropriate pipeline based on the contents of the Module by default. This should be made even more flexible in future (e.g., mixing the different input conversions), but removes a footgun and improves user experience.

This does not cover every config (and this change does not removed explicitly specified input conversion flags in tests), but should cover reasonable set. The cost is an extra traversal and in future this could be made even more dynamic (e.g., handling combinations of these dialects). Open to better naming suggestions for conversion type.

Related to #10297.

jpienaar committed:
Reduce need to specify input pipeline.

Utilize dynamic pipeline to construct appropriate pipeline based on the
contents of the Module by default. This should be made even more flexible in
future (e.g., mixing the different input conversions), but removes a
footgun/improves user experience for common case.

Related to #10297.

jpienaar committed:
Fix changes dropped cmake side

jpienaar committed:
Format

MaheshRavishankar commented on compiler/src/iree/compiler/InputConversion/Common/CMakeLists.txt:
```diff
@@ -8,8 +8,32 @@
 
+# Enable input dialects based on options.
+set(IREE_INPUT_DEPS "")
+if(IREE_INPUT_MHLO)
```
I think this file is supposed to be auto-generated.... I dont think the bazel changes + bazel_to_cmake is consistent with the changes in this file.

jpienaar replied:
Changed header so that it isn't autogenerated (this is following Pipeline.cpp structure).

ScottTodd replied:
These changes do make sense, though we could restructure a bit if we want to keep it autogenerated. (I'm fine with just leveraging autogeneration where it helps and making manual changes like this as needed)

MaheshRavishankar commented on compiler/src/iree/compiler/InputConversion/Common/InputConversionPipeline.cpp:
```diff
@@ -0,0 +1,146 @@
+  const Dialect* mhloDialect = nullptr;
+  const Dialect* tosaDialect = nullptr;
+  const Dialect* tmTensorDialect = nullptr;
+#ifdef IREE_HAVE_MHLO_INPUT
```
I am not sure having `ifdef`s in the code is the way we want to support this.....

jpienaar replied:
It is how we are currently supporting these. I could switch to string based version (`getLoadedDialect(StringRef name)`) and then just use ifdefs for includes.

ScottTodd replied:
Yep, the defines let us make each input type (and associated submodule, build, test, etc. deps) optional. `getLoadedDialect` is an interesting idea... that would let us centralize the defines to make the rest of the code a little cleaner.

jpienaar committed:
Lookup dialect by name & update cmake header

jpienaar committed:
Update bazel deps

jpienaar committed:
Rename to auto

Easier to understand than analysis. Also split out into separate library, was not noticing the usage as was doing cmake builds.

jpienaar committed:
Update bazel deps (lint script not working on mac and bazel is also not playing along)

jpienaar committed:
Add dialect deps in bazel that were transitive but not on lib

ScottTodd commented on compiler/src/iree/compiler/InputConversion/Common/InputConversionPipeline.cpp:
```diff
@@ -0,0 +1,132 @@
+
+  // tm_tensor
+  bool hasTmTensor = false;
+};
```
Could use a bitfield instead of a struct of `bool`s here. Not sure what LLVM/MLIR style is for that, but on the runtime side we have a bunch like those in https://github.com/openxla/iree/blob/main/runtime/src/iree/hal/buffer.h

jpienaar replied:
Runtime this makes sense, here I feel it doesn't pay for itself. But could follow up post measurement.

ScottTodd commented on compiler/src/iree/compiler/InputConversion/Common/InputConversionPipeline.cpp:
```diff
@@ -0,0 +1,132 @@
+    else
+      MHLO::buildMHLOInputConversionPassPipeline(pm);
+  }
+  if (features.hasTOSA) buildTOSAInputConversionPassPipeline(pm);
```
nit: add braces?

ScottTodd commented on compiler/src/iree/compiler/InputConversion/Common/Passes.td:
```diff
@@ -45,4 +45,14 @@ def TopLevelSCFToCFG :
+    conversion to run.
+  }];
+  let constructor = "mlir::iree_compiler::createInputConversionPipelinePass()";
+}
```
grammar

```suggestion
def InputConversionPipeline :
    Pass<"iree-input-conversion", "ModuleOp"> {
  let summary = "Analyzes and runs appropriate input pipeline.";
  let description = [{
    Analyzes the input program to determine which input pipeline, or part of input
    conversion to run, then runs that conversion.
  }];
  let constructor = "mlir::iree_compiler::createInputConversionPipelinePass()";
}
```

ScottTodd commented on compiler/src/iree/compiler/Pipelines/Pipelines.cpp:
```diff
@@ -49,6 +49,9 @@ void buildIREEVMTransformPassPipeline(
     case InputDialectOptions::Type::none:
       break;
+    case InputDialectOptions::Type::auto_detect:
+      passManager.addPass(createInputConversionPipelinePass());
```
Maybe this pass should have "auto"/"inferred" in the name?

jpienaar committed:
Address review comments
